pennylane/grouping/pauli.py

```diff
@@ -0,0 +1,267 @@
+    ...    print(p)
+    ...
+    Identity(wires=[0])
+    PauliZ(wires=[2])
+    PauliZ(wires=[1])
+    PauliZ(wires=[1]) @ PauliZ(wires=[2])
+    PauliZ(wires=[0])
+    PauliZ(wires=[0]) @ PauliZ(wires=[2])
+    PauliZ(wires=[0]) @ PauliZ(wires=[1])
+    ...
+
+    The full Pauli group can then be obtained like so:
+
+    >>> full_pg = list(pauli_group(n_qubits))
+
+    The group can also be created using a custom wire map (if no map is
+    specified, a default map of label :math:`i` to wire ``i`` as in the example
+    above will be created).
+
+    >>> wire_map = {'a' : 0, 'b' : 1, 'c' : 2}
+    >>> for p in pauli_group(n_qubits, wire_map=wire_map):
+    ...     print(p)
+    ...
+    Identity(wires=['a'])
+    PauliZ(wires=['c'])
+    PauliZ(wires=['b'])
+    PauliZ(wires=['b']) @ PauliZ(wires=['c'])
+    PauliZ(wires=['a'])
+    PauliZ(wires=['a']) @ PauliZ(wires=['c'])
+    PauliZ(wires=['a']) @ PauliZ(wires=['b'])
+    ...
+
+    """
+    if not isinstance(n_qubits, int):
+        raise TypeError("Must specify an integer number of qubits construct the Pauli group.")
+
+    if n_qubits <= 0:
+        raise ValueError("Number of qubits must be at least 1 to construct Pauli group.")
+
+    return _pauli_group_generator(n_qubits, wire_map=wire_map)
+
+
+def pauli_mult(pauli_1, pauli_2, wire_map=None):
+    """Multiply two Pauli words together and return the product as a Pauli word.
+
+    Two Pauli operations can be multiplied together by taking the additive
+    OR of their binary symplectic representations.
+
+    Args:
+        pauli_1 (.Operation): A Pauli word.
+        pauli_2 (.Operation): A Pauli word to multiply with the first one.
+        wire_map (dict[Union[str, int], int]): dictionary containing all wire labels used in the Pauli
+            word as keys, and unique integer labels as their values. If no wire map is
+            provided, the map will be constructed from the set of wires acted on
+            by the input Pauli words.
+
+    Returns:
+        .Operation: The product of pauli_1 and pauli_2 as a Pauli word
+        (ignoring the global phase).
+
+    **Example**
+
+    This function enables multiplication of Pauli group elements at the level of
+    Pauli words, rather than matrices. For example,
+
+    >>> from pennylane.grouping import pauli_mult
+    >>> pauli_1 = qml.PauliX(0) @ qml.PauliZ(1)
+    >>> pauli_2 = qml.PauliY(0) @ qml.PauliZ(1)
+    >>> product = pauli_mult(pauli_1, pauli_2)
+    >>> print(product)
+    PauliZ(wires=[0])
+    """
+
+    # If no wire map is specified, generate one from the union of wires
+    # in both Paulis.
+    if wire_map is None:
+        wire_labels = Wires.all_wires([pauli_1.wires, pauli_2.wires]).labels
+        wire_map = {label: i for i, label in enumerate(wire_labels)}
+
+    # Check if pauli_1 and pauli_2 are the same; if so, the result is the Identity
+    if are_identical_pauli_words(pauli_1, pauli_2):
+        first_wire = list(wire_map.keys())[0]
```

Comment:
I think this should be fine, though it will spit out an `Identity` that is not necessarily on the first qubit in the set. For example, with a `wire_map = {0:0, 1:1, 2:2}`, asking for `pauli_mult(PauliX(1) @ PauliY(2), PauliX(1) @ PauliY(2))` would return `Identity(1)` instead of `Identity(0)`. But since it's the identity, I don't foresee any issues.